backend/src/database/state-migrations/migrations.ts

```diff
@@ -1,52 +1,42 @@
-import type { UUID } from 'digital-fuesim-manv-shared';
+import type {
+    ExerciseState,
+    StateExport,
+    UUID,
+} from 'digital-fuesim-manv-shared';
 import type { EntityManager } from 'typeorm';
-import type { ExerciseWrapper } from '../../exercise/exercise-wrapper';
 import { RestoreError } from '../../utils/restore-error';
+import { ActionWrapperEntity } from '../entities/action-wrapper.entity';
+import { ExerciseWrapperEntity } from '../entities/exercise-wrapper.entity';
+import { impossibleMigration } from './impossible-migration';
 
 /**
- * Such a function MUST update the initial state of the exercise with the provided {@link exerciseId} as well as every action associated with it from its current state version to the next version in a way that they are valid states/actions.
- * It MAY throw a {@link RestoreError} in a case where upgrading is impossible and a terminal incompatibility with older exercises is necessary.
- * It MUST update the respective updates to the exercise and its associated objects in the database.
- * All database interaction MUST use the provided {@link EntityManager}.
+ * Such a function gets the already migrated initial state of the exercise and an array of all actions (not yet migrated).
+ * It is expected that afterwards the actions in the provided array are migrated.
+ * It is not allowed to modify the order of the actions, to add an action or to remove an action.
+ * To indicate that an action should be removed it can be replaced by `null`.
+ * It may throw a {@link RestoreError} when a migration is not possible.
  */
-export type DbMigrationFunction = (
-    entityManager: EntityManager,
-    exerciseId: UUID
-) => Promise<void>;
+type MigrateActionsFunction = (
+    initialState: object,
+    actions: (object | null)[]
+) => void;
 
 /**
- * Such a function MUST update the initial state of the provided {@link exerciseWrapper} as well as every action associated with it from its current state version to the next version in a way that they are valid states/actions.
- * It MAY throw a {@link RestoreError} in a case where upgrading is impossible and a terminal incompatibility with older exercises is necessary.
- * It MUST NOT use the database.
+ * Such a function gets the not yet migrated state and is expected to return a migrated version of it.
+ * It may throw a {@link RestoreError} when a migration is not possible.
  */
-export type InMemoryMigrationFunction = (
-    exerciseWrapper: ExerciseWrapper
-) => Promise<ExerciseWrapper>;
+type MigrateStateFunction = (state: object) => object;
 
-export interface MigrationFunctions {
-    database: DbMigrationFunction;
-    inMemory: InMemoryMigrationFunction;
+export interface MigrationSpecification {
```

Comment:
```suggestion
export interface Migration {
```

I would still only call this a `Migration`. An interface is always some sort of Specification and the inheritance naming problem doesn't apply anymore because they are now called `Migrate...Function`.

You can then use:

```ts
const migrations: Migration[];
const migration: Migration;
const MigrateStateFunctions: MigrateStateFunction[];
...
```